File: examples/python/model-qa.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import argparse
import json
import time

import onnxruntime_genai as og
from common import apply_chat_template, get_config, get_generator_params_args, get_guidance, get_guidance_args, get_search_options
# og.set_log_options(enabled=True, model_input_values=True, model_output_values=True)

def main(args):
    if args.verbose:
        print("Loading model...")

    # Create model
    config = get_config(args.model_path, args.execution_provider)
    model = og.Model(config)
    if args.verbose:
        print("Model loaded")

    # Create tokenizer
    tokenizer = og.Tokenizer(model)
    tokenizer_stream = tokenizer.create_stream()
    if args.verbose:
        print("Tokenizer created")

    # Get search options for generator params
    search_options = get_search_options(args)

    # Create running list of messages
    input_list = [
        {"role": "system", "content": args.system_prompt},
    ]

    # Get guidance info if requested
    if args.response_format != "":
        print("Make sure your tool call start id and tool call end id are marked as special in tokenizer.json")
        guidance_type, guidance_data, tools = get_guidance(
            response_format=args.response_format,
            filepath=args.tools_file,
            text_output=args.text_output,
            tool_output=args.tool_output,
            tool_call_start=args.tool_call_start,
            tool_call_end=args.tool_call_end,
        )
        input_list[0]["tools"] = tools

    # Keep track of timings if requested
    if args.timings:
        started_timestamp = 0
        first_token_timestamp = 0

    # Keep asking for input prompts in a loop
    while True:
        # Get user prompt
        if args.input_prompt:
            text = args.input_prompt
        else:
            text = input("Prompt (Use quit() to exit): ")
        if not text:
            print("Error, input cannot be empty")
            continue

        if text == "quit()":
            break

        # Add user message to list of messages
        input_list.append({"role": "user", "content": text})
        messages = json.dumps(input_list)
    
        if args.timings:
            started_timestamp = time.time()

        # Initialize generator params
        params = og.GeneratorParams(model)
        params.set_search_options(**search_options)
        if args.verbose:
            print(f"GeneratorParams created: {search_options}")

        # Initialize guidance info
        if args.response_format != "":
            params.set_guidance(guidance_type, guidance_data)
            if args.verbose:
                print()
                print(f"Guidance type is: {guidance_type}")
                print(f"Guidance data is: \n{guidance_data}")
                print()

        # Create generator
        generator = og.Generator(model, params)
        if args.verbose:
            print("Generator created")

        # Apply chat template
        try:
            prompt = apply_chat_template(model_path=args.model_path, tokenizer=tokenizer, messages=messages, tools=tools, add_generation_prompt=True)
        except:
            prompt = text
        if args.verbose:
            print(f"Prompt: {prompt}")

        # Encode combined system + user prompt and append tokens to model
        input_tokens = tokenizer.encode(prompt)
        generator.append_tokens(input_tokens)

        if args.verbose:
            print("Running generation loop ...")
        if args.timings:
            first = True
            new_tokens = []

        print()
        print("Output: ", end="", flush=True)

        # Run generation loop
        try:
            while True:
                generator.generate_next_token()
                if args.timings:
                    if first:
                        first_token_timestamp = time.time()
                        first = False

                if generator.is_done():
                    break

                new_token = generator.get_next_tokens()[0]
                print(tokenizer_stream.decode(new_token), end="", flush=True)
                if args.timings:
                    new_tokens.append(new_token)
        except KeyboardInterrupt:
            print("  --control+c pressed, aborting generation--")
        print()
        print()

        # Delete the generator to free the captured graph for the next generator (if graph capture is enabled)
        del generator

        # Remove user message from list of messages
        input_list.pop()

        if args.timings:
            prompt_time = first_token_timestamp - started_timestamp
            run_time = time.time() - first_token_timestamp
            print(
                f"Prompt length: {len(input_tokens)}, New tokens: {len(new_tokens)}, Time to first: {(prompt_time):.2f}s, Prompt tokens per second: {len(input_tokens) / prompt_time:.2f} tps, New tokens per second: {len(new_tokens) / run_time:.2f} tps"
            )

        # If input prompt is provided, it will just run the model for the input prompt and exit
        if args.input_prompt:
            break


if __name__ == "__main__":
    parser = argparse.ArgumentParser(argument_default=argparse.SUPPRESS, description="End-to-end AI question/answer example for ORT GenAI")
    parser.add_argument('-m', '--model_path', type=str, required=True, help='ONNX model folder path (must contain genai_config.json and model.onnx)')
    parser.add_argument('-e', '--execution_provider', type=str, required=False, default='follow_config', choices=["cpu", "cuda", "dml", "follow_config"], help="Execution provider to run the ONNX Runtime session with. Defaults to follow_config that uses the execution provider listed in the genai_config.json instead.")
    parser.add_argument('-v', '--verbose', action='store_true', default=False, help='Print verbose output and timing information. Defaults to false')
    parser.add_argument('-g', '--timings', action='store_true', default=False, help='Print timing information for each generation step. Defaults to false')
    parser.add_argument('-sp', '--system_prompt', type=str, default='You are a helpful AI assistant.', help='System prompt to use for the prompt.')
    parser.add_argument('-ip', '--input_prompt', type=str, default='', help='Input prompt. If provided, it will just run the prompt and exit.')

    get_generator_params_args(parser)
    get_guidance_args(parser)

    args = parser.parse_args()
    main(args)

File: examples/c/src/model_qa.cpp
// -----------------------------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
//
// C++ API Example for Model Question-Answering
// This example demonstrates how to use the C++ API of the ONNX Runtime GenAI library
// to perform model question-answering tasks. It includes functionalities to create a model,
// tokenizer, and generator, and to handle user input for generating responses based on prompts.
// -----------------------------------------------------------------------------------------------

#include <csignal>
#include <iomanip>
#include <string>

#include "common.h"

OgaGenerator* g_generator = nullptr;

void TerminateGeneration(int signum) {
  if (g_generator == nullptr) {
    return;
  }
  g_generator->SetRuntimeOption("terminate_session", "1");
}

void CXX_API(
  GeneratorParamsArgs& generator_params_args,
  GuidanceArgs& guidance_args,
  const std::string& model_path,
  const std::string& ep,
  const std::string& system_prompt,
  bool verbose,
  bool interactive
) {
  if (verbose) std::cout << "Creating config..." << std::endl;
  std::unordered_map<std::string, std::string> ep_options;
  auto config = GetConfig(model_path, ep, ep_options, generator_params_args);

  if (verbose) std::cout << "Creating model..." << std::endl;
  auto model = OgaModel::Create(*config);

  if (verbose) std::cout << "Creating tokenizer..." << std::endl;
  auto tokenizer = OgaTokenizer::Create(*model);
  auto tokenizer_stream = OgaTokenizerStream::Create(*tokenizer);

  // Create running list of messages
  std::vector<nlohmann::ordered_json> input_list;
  nlohmann::ordered_json system_message = nlohmann::ordered_json{ {"role", "system"}, {"content", system_prompt} };
  input_list.push_back(system_message);

  // Get and set guidance info if requested
  std::string guidance_type, guidance_data, tools;
  if (!guidance_args.response_format.empty()) {
    std::cout << "Make sure your tool call start id and tool call end id are marked as special in tokenizer.json" << std::endl;
    std::tie(guidance_type, guidance_data, tools) = GetGuidance(
      guidance_args.response_format,
      guidance_args.tools_file,
      "",  // tools_str
      nullptr,  // tools
      guidance_args.text_output, 
      guidance_args.tool_output,
      guidance_args.tool_call_start,
      guidance_args.tool_call_end
    );
    
    input_list[0]["tools"] = tools;
  }

  // Keep asking for input prompts in a loop
  while (true) {
    // Get user prompt
    std::string text;
    std::cout << "Prompt (Use quit() to exit):" << std::endl;
    // Clear Any cin error flags because of SIGINT
    std::cin.clear();
    std::getline(std::cin, text);

    if (text.empty()) {
      std::cout << "Empty input. Please enter a valid prompt." << std::endl;
      continue;  // Skip to the next iteration if input is empty
    } else if (text == "quit()") {
      break;  // Exit the loop
    }

    signal(SIGINT, TerminateGeneration);

    // Add user message to list of messages
    nlohmann::ordered_json user_message = nlohmann::ordered_json{ {"role", "user"}, {"content", text} };
    input_list.push_back(user_message);
    nlohmann::ordered_json j = input_list;
    std::string messages = j.dump();

    // Start timings
    bool is_first_token = true;
    Timing timing;
    timing.RecordStartTimestamp();

    // Initialize generator params
    auto params = OgaGeneratorParams::Create(*model);
    SetSearchOptions(*params, generator_params_args, verbose);

    // Initialize guidance info
    if (!guidance_args.response_format.empty()) {
      params->SetGuidance(guidance_type.c_str(), guidance_data.c_str());
      if (verbose) {
        std::cout << std::endl;
        std::cout << "Guidance type is: " << guidance_type << std::endl;
        std::cout << "Guidance data is: \n" << guidance_data << std::endl;
        std::cout << std::endl;
      }
    }

    // Create generator
    auto generator = OgaGenerator::Create(*model, *params);
    g_generator = generator.get();  // Store the current generator for termination
    if (verbose) std::cout << "Generator created" << std::endl;

    // Apply chat template
    std::string prompt;
    try {
      bool add_generation_prompt = true;
      prompt = ApplyChatTemplate(model_path, *tokenizer, messages, add_generation_prompt, tools);
    }
    catch (...) {
      prompt = text;
    }
    if (verbose) std::cout << "Prompt: " << prompt << "\n" << std::endl;

    // Encode combined system + user prompt and append tokens to model
    auto sequences = OgaSequences::Create();
    tokenizer->Encode(prompt.c_str(), *sequences);
    generator->AppendTokenSequences(*sequences);

    // Run generation loop
    if (verbose) std::cout << "Running generation loop..." << std::endl;
    std::cout << std::endl;
    std::cout << "Output: ";
    try {
      while (true) {
        generator->GenerateNextToken();

        if (is_first_token) {
          timing.RecordFirstTokenTimestamp();
          is_first_token = false;
        }

        if (generator->IsDone()) {
          break;
        }

        const auto num_tokens = generator->GetSequenceCount(0);
        const auto new_token = generator->GetSequenceData(0)[num_tokens - 1];
        std::cout << tokenizer_stream->Decode(new_token) << std::flush;
      }
    } catch (const std::exception& e) {
      std::cout << "\n" << "Terminating generation: " << e.what() << std::endl;
    }
    timing.RecordEndTimestamp();

    // Clear the generator after use
    g_generator = nullptr;

    // Remove user message from list of messages
    input_list.pop_back();

    const int prompt_tokens_length = sequences->SequenceCount(0);
    const int new_tokens_length = generator->GetSequenceCount(0) - prompt_tokens_length;
    timing.Log(prompt_tokens_length, new_tokens_length);

    std::cout << "\n\n" << std::endl;
  }
}

int main(int argc, char** argv) {
  // Get command-line args
  GeneratorParamsArgs generator_params_args;
  GuidanceArgs guidance_args;
  std::string model_path, ep = "follow_config", system_prompt = "You are a helpful AI assistant.";
  bool verbose = false, interactive = false, rewind = true;
  if (!ParseArgs(argc, argv, generator_params_args, guidance_args, model_path, ep, system_prompt, verbose, interactive, rewind)) {
    return -1;
  }

  // Responsible for cleaning up the library during shutdown
  OgaHandle handle;

  std::cout << "--------------------------" << std::endl;
  std::cout << "Hello, ORT GenAI Model-QA!" << std::endl;
  std::cout << "--------------------------" << std::endl;

  std::cout << "Model path: " << model_path << std::endl;
  std::cout << "Execution provider: " << ep << std::endl;
  std::cout << "System prompt: " << system_prompt << std::endl;
  std::cout << "Verbose: " << verbose << std::endl;
  std::cout << "Interactive: " << interactive << std::endl;
  std::cout << "--------------------------" << std::endl;
  std::cout << std::endl;

  try {
    CXX_API(generator_params_args, guidance_args, model_path, ep, system_prompt, verbose, interactive);
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return -1;
  }

  return 0;
}